docs/src/appendix/mast.md

```diff
@@ -0,0 +1,526 @@
+
+External nodes do not sacrifice any properties of the MAST, because they preserve the integrity of the MAST hash, and have the exact same semantics as if the referenced node was part of the same `MastForest`. The primary difference is that if the referenced node is not available at runtime for some reason, then any attempt to execute that node will fail.
+
+### Binary Format
+
+We can now cover the binary format for the structure we've outlined above. The two are independent of one another, but are designed such that serializing/deserializing between them is simple and efficient. First, we will cover the specification for the binary format itself, and then we will cover the specific serialization/deserialization procedures for the in-memory structure we've defined.
+
+#### Overview
+
+To make the specification easier to understand at a glance, we're going to use Rust code to define the layout, with the following modified syntax/semantics:
+
+* All structs are assumed to use `#[repr(C)]` unless a different `#[repr]` is given, or modifications are otherwise noted
+* We relax the restriction on dynamically-sized fields in the middle of structs,
+as otherwise it would not be possible to express the format in Rust syntax
+* We allow specifying the size of array-typed fields using self-referential expressions, as long as the referenced fields in the expression precede the array-typed field. For example, `items: [T; self.len]` is valid if the `len` field preceds the `items` field in the struct definition.
+
+#### Variable-Width Integers
+
+First, let's introduce a type that we will use throughout the rest of this spec, `VInt64`. It represents a variable-width, unsigned, little-endian 64-bit integer. It is encoded as one or more bytes, where each byte has 7 bits to represent the value, with the last bit used to signal when another byte is needed to represent the encoded value. In this case, unless otherwise noted, we treat it as an error if a `VInt64`-encoded integer would not fit in the integral type given.
+
+See [vint64](https://docs.rs/vint64/latest/vint64/) for more details on the representation/encoding.
+
+#### Encoding
+
+The following is the binary encoding of a `MastForest` using our specification language as described above:
+
+```rust
+#[repr(C)]
+pub struct EncodedMastForest {
+    /// The magic string `b"MAST\0"`, for detecting that a file
+    /// is binary-encoded MAST
+    magic: [u8; 5],
+    /// The format version, initially all zeros
+    ///
+    /// If future modifications are made to this format, the version
+    /// should be incremented by 1. A version of 0b111 is reserved
+    /// for future extensions that require extending the version
+    /// field itself, but should be considered invalid for now.
+    version: [u8; 3],
```

Comment:
One other question re `version`: would this implicitly encode the VM version as well? Specifically, let's say we modify the instruction set in some way - would the format version be incremented? If not, would it make sense to add something like `miden_version` field here (instead of it being in the package definition)?